### PR TITLE
Correctly handle errors for incoming requests

### PR DESCRIPTION
**Pull request by jcorbin**
- firstly, finish handling a server op when we send an error frame
- secondly, hookup req error events to send an error frame

r @kriskowal @Raynos

### CODE REVIEW

**Comment by jcorbin** [comment]
Depends on #370, review it first.

File: node/connection_base.js [hunk 1]
```diff
@@ -269,12 +269,19 @@ TChannelConnectionBase.prototype.handleCallRequest = function handleCallRequest(
...
     process.nextTick(runHandler);
 
     if (req.span) {
         req.span.endpoint.serviceName = self.channel.serviceName;
     }
 
+    function onReqError(err) {
+        if (!op.res) buildResponse();
+        var errName = err.name || err.constructor.name;
+        op.res.sendError('UnexpectedError', errName + ': ' + err.message);
```
Comment (kriskowal) [inline node/connection_base.js]: I must have misread the source. This is not supposed to be `'unexpected error'`?

File: node/connection_base.js [hunk 2]
```diff
@@ -285,10 +292,13 @@ TChannelConnectionBase.prototype.handleCallRequest = function handleCallRequest(
...
     function buildResponse(options) {
         if (op.res && op.res.state !== OutgoingResponse.States.Initial) {
-            throw new Error('response already built and started'); // TODO: typed error
+            throw errors.ResponseAlreadyStarted({
+                state: op.res.state
+            });
         }
         op.res = self.buildOutgoingResponse(req, options);
         op.res.on('finish', opDone);
+        op.res.on('errored', opDone);
```
Comment (kriskowal) [inline node/connection_base.js]: Avoiding a collision with 'error'?

File: node/connection_base.js (hunk 1 above)
Comment (jcorbin) [inline node/connection_base.js]: why not? we don't know anything about the error at this point, in the future we'll add some typed cases.

e.g. I've got one in queue to add timeout -> timed out coupling

File: node/connection_base.js (hunk 2 above)
Comment (jcorbin) [inline node/connection_base.js]: correct, it's simply noting that the response is over due to an error response... bit hack perhaps

Comment (kriskowal) [inline node/connection_base.js]: :heavy_check_mark:

File: node/connection_base.js (hunk 1 above)
Comment (kriskowal) [inline node/connection_base.js]: Over the shoulder, 'UnexpectedError' it is. :heavy_check_mark:

**Comment by kriskowal** [comment]
:ok:

**Pull request merged**